libs/vm/src/sharded_state.cpp

```diff
@@ -0,0 +1,132 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "vm/sharded_state.hpp"
+
+namespace fetch {
+namespace vm {
+namespace {
+class ShardedState : public IShardedState
+{
+public:
+  // Construction / Destruction
+  ShardedState(VM *vm, TypeId type_id, Ptr<String> const &name, TypeId value_type)
+    : IShardedState(vm, type_id, name, value_type)
+  {}
+
+protected:
+  using ByteBuffer = std::vector<uint8_t>;
+
+  TemplateParameter1 GetIndexedValue(Ptr<String> const &key) override
+  {
+    return GetIndexedValueInternal(key);
+  }
+
+  void SetIndexedValue(Ptr<String> const &key, TemplateParameter1 const &value) override
+  {
+    SetIndexedValueInternal(key, value);
+  }
+
+  TemplateParameter1 GetIndexedValue(Ptr<Address> const &key) override
+  {
+    if (!key)
+    {
+      RuntimeError("Index is null reference.");
+      return {};
+    }
+    return GetIndexedValueInternal(key->AsString());
+  }
+
+  void SetIndexedValue(Ptr<Address> const &key, TemplateParameter1 const &value) override
+  {
+    if (!key)
+    {
+      RuntimeError("Index is null reference.");
+    }
+    SetIndexedValueInternal(key->AsString(), value);
+  }
+
```

Comment:
String key could be null too, no?

Reply:
Yes, but it is checked for null in side of `SetIndexedValueInternal(...)` by `ComposeFullKey(...)` method.